### DISCUSSION TITLE
Retrieve a Quantity Unit from given BaseUnits

### DISCUSSION THREAD

**Discussion by devel0** [angularsen/UnitsNet#1014] [Q&A] [answered]
I seen not all units in json have BaseUnits, for example [VolumeUnit.CubicMillimeter](https://github.com/angularsen/UnitsNet/blob/2c3cb93cffd453a296a64ed0eadc5d02e35a4836/Common/UnitDefinitions/Volume.json#L113-L128) lacks for BaseUnits declaration while other for example [VolumeUnit.CubicMeter](https://github.com/angularsen/UnitsNet/blob/2c3cb93cffd453a296a64ed0eadc5d02e35a4836/Common/UnitDefinitions/Volume.json#L26-L45) does.

I understand that this is justified for unit like [VolumeUnit.MetricCup](https://github.com/angularsen/UnitsNet/blob/2c3cb93cffd453a296a64ed0eadc5d02e35a4836/Common/UnitDefinitions/Volume.json#L376-L387) because there isn't a proper defined base LengthUnit such that [L^3](https://github.com/angularsen/UnitsNet/blob/2c3cb93cffd453a296a64ed0eadc5d02e35a4836/Common/UnitDefinitions/Volume.json#L5-L7) results in 1 cup.

Btw I would to know if the lacks of BaseUnits for already mentioned CubicMillimeter is because simply json isn't yet stable for BaseUnits description on all derived measure units directly derived from base units or if there are some other reasons.

I interested in this because actually I would to write a function ComputeVolumeUnit from given LengthUnit like following:

```cs
public static VolumeUnit ComputeVolumeUnit(LengthUnit lengthUnit)
{
    var res = Volume.Info.UnitInfos.FirstOrDefault(w => w.BaseUnits.Length == lengthUnit)?.Value;

    if (!res.HasValue)
        throw new InternalError($"not implemented Volume from length unit = {lengthUnit}");

    return res.Value;
}
```

while actually, cause the problem exposed, I have to write this:

```cs
public static VolumeUnit ComputeVolumeUnit(LengthUnit lengthUnit)
{
    VolumeUnit res;            

    switch (lengthUnit)
    {
        case LengthUnit.Meter: res = VolumeUnit.CubicMeter; break;
        case LengthUnit.Centimeter: res = VolumeUnit.CubicCentimeter; break;
        case LengthUnit.Millimeter: res = VolumeUnit.CubicMillimeter; break;
        case LengthUnit.Foot: res = VolumeUnit.CubicFoot; break;
        case LengthUnit.Inch: res = VolumeUnit.CubicInch; break;

        default: throw new InternalError($"not implemented Volume from length unit = {lengthUnit}");
    }

    return res;
}
```

the same manner a ComputeDensityVolumeUnit(MassUnit massUnit, LengthUnit lenghUnit) would more longer than a version using the search through UnitInfos.

**Comment by angularsen** [comment, marked as answer]
Hi, the idea with BaseUnits is to describe the 7 SI base units from which all other SI units can be derived.
https://en.wikipedia.org/wiki/SI_base_unit

I believe it was introduced as an experiment to be able to configure default units in an application (meter, centimeter, foot). If Length was configured to centimeters, then derived units like area and volume would also automatically be square/cubic centimeters. 

It also somewhat helps you represent that Area * Area / Length = Volume by defining arithmetic for the dimensions, but you currently can't multiply two `IQuantity` of unknown quantity type and unit and get a result back.

The problem is that we have many non-SI units. In addition, multiple units map to the same 7 base units. The usefulness of this base units feature is debatable and it is often confusing for contributors what base units to define.

In the upcoming v5 release, it is considered to be removed until the idea is more mature and supports non-SI units.
https://github.com/angularsen/UnitsNet/pull/982#issuecomment-957008923


So long story, but yes, I believe you have to write the piece of code you already did as it is now.
If you see a better design for this whole unit system and base unit thing, then we are open to ideas. The momentum behind it has stagnated and no one is currently working on it.

**Reply by mwagnerEE** [reply]
If you're open to ideas, I think this could be a very useful thing. I think a big use case of UnitsNet could be as a backbone for scientific calculations. I'm trying to use it for some power calculations but unless there is predefined operator I cant just multiply any two arbitrary quantities together. It might not make sense to multiply `Power` * `Power`, but if that is the first step of a formula temporarily have a quantity with units Watts^2 is okay.

My idea would be to have a `NonstandardQuantity`. So if you multiply any arbitrary quantities, there will be a `Quantity.FromBaseUnits` method that checks to see if a standard quantity with those base units exists and returns that. However if it doesnt exist, a `NonstandardQuantity` is returned which can then be multiplied again until the calculation is done and you either end up with a standard quantity or a non-standard one. `NonstandardQuantity` doesn't need to be a new special type, you could make a base struct `Quantity` that the other quantities inherit as they are just special cases.

I like the base units but I understand not every quantity uses only SI units. Perhaps if instead of a standard array of SI units, `IQuantity.Dimensions` was a `Dictionary<Enum, int>` that could allow arbitrary units and keep track of their exponents adding or subtracting as calculations are performed. Any two quantities could be multiplied or divided but an exception could be thrown if adding or subtracting two quantities with different units?

**Reply by angularsen** [reply]
Yes, I think it would be interesting to support dynamic quantities to support arithmetic.
Someone with a real need for it and a concrete use case would be best able make it happen. I don't have the need for it, nor the time 😅 I'm mostly concerned about not adding something that won't be used.

As for SI units, I think I've come to learn that just about all units can be expressed from SI base units and I'm not sure if there is a need for any other unit system than SI. @tmilnthorp has a better grasp of it than me.

Some more info here:
https://github.com/angularsen/UnitsNet/pull/982#issuecomment-1040798580

**Reply by mwagnerEE** [reply]
I was able to get this working, although when I first pitched that idea, I forgot that you cant inherit from a struct. So I ended up cannibalizing  the whole thing to use classes, which is way beyond a feature request lol. Although I believe it would be possible with structs, it would not be straightforward.

I just got it working today but it's able to do pretty much what I described where it will return a defined quantity if it can math the units otherwise it will return a `QuantityBase` which is a class that everything derives from.

Now I can do stuff like this:
```cs
ElectricCurrent i = new ElectricCurrent(5, ElectricCurrentUnit.Ampere);
ElectricResistance r = new ElectricResistance(10, ElectricResistanceUnit.Ohm);
ApparentPower s = (ApparentPower)(i.Pow(2) * r);
```

Which prevents the need for creating an extra method
```cs
ApparentPower.FromCurrentAndResistance(ElectricCurrent i, ElectricResistance r);
```

I'd love to maybe share it but it is quite dirty. As a proof of concept, I didn't really follow the coding guidelines lol.

**Reply by mwagnerEE** [reply]
I cleaned it up somewhat and created a branch if you wanna check it out. [https://github.com/mwagnerEE/UnitsNet](url)
Let me know if you want me to change how I uploaded it or anything. I tried to make sure it was a branch so it didn't look like a repost and I changed the readme to make it clear its a "demo".

**Comment by tmilnthorp** [comment]
It is simply due to it missing in the json files. We should add where it's missing.
